File: examples/web_api_client.py
#!/usr/bin/env python3
"""
Example Python client for slopdrop Web API
Demonstrates how to interact with slopdrop programmatically

Usage:
    python3 examples/web_api_client.py
"""

import requests
import json
import sys

class SlopdropClient:
    def __init__(self, base_url="http://127.0.0.1:8080"):
        self.base_url = base_url
        self.session = requests.Session()

    def eval(self, code, is_admin=False):
        """Evaluate TCL code"""
        response = self.session.post(
            f"{self.base_url}/api/eval",
            json={
                "code": code,
                "is_admin": is_admin
            }
        )
        response.raise_for_status()
        return response.json()

    def more(self):
        """Get more paginated output"""
        response = self.session.get(f"{self.base_url}/api/more")
        response.raise_for_status()
        return response.json()

    def history(self, limit=10):
        """Get git commit history"""
        response = self.session.get(
            f"{self.base_url}/api/history",
            params={"limit": limit}
        )
        response.raise_for_status()
        return response.json()

    def rollback(self, commit_hash):
        """Rollback to specific commit"""
        response = self.session.post(
            f"{self.base_url}/api/rollback",
            json={"commit_hash": commit_hash}
        )
        response.raise_for_status()
        return response.json()

    def health(self):
        """Check server health"""
        response = self.session.get(f"{self.base_url}/api/health")
        response.raise_for_status()
        return response.json()


def main():
    # Create client
    client = SlopdropClient()

    # Check health
    print("Checking server health...")
    health = client.health()
    print(f"✓ Server is {health['status']}\n")

    # Simple evaluation
    print("Evaluating: expr {1 + 1}")
    result = client.eval("expr {1 + 1}")
    print(f"Result: {result['output']}")
    print(f"Error: {result['is_error']}\n")

    # Set a variable
    print("Evaluating: set myvar \"Hello from Python!\"")
    result = client.eval("set myvar \"Hello from Python!\"")
    print(f"Result: {result['output']}\n")

    # Read the variable
    print("Evaluating: set myvar")
    result = client.eval("set myvar")
    print(f"Result: {result['output']}\n")

    # Define a procedure
    print("Defining factorial procedure...")
    factorial_code = """
proc factorial {n} {
    if {$n <= 1} {
        return 1
    }
    expr {$n * [factorial [expr {$n - 1}]]}
}
    """.strip()
    result = client.eval(factorial_code, is_admin=True)
    print(f"Result: {result['output']}\n")

    # Call the procedure
    print("Evaluating: factorial 5")
    result = client.eval("factorial 5")
    print(f"Result: {result['output']}\n")

    # Get history
    print("Getting git history...")
    history = client.history(5)
    print("Recent commits:")
    for commit in history['history']:
        print(f"  {commit['commit_id'][:8]} - {commit['author']} - {commit['message']}")
    print()

    # Test pagination with large output
    print("Testing pagination with large output...")
    result = client.eval("for {set i 0} {$i < 50} {incr i} { puts \"Line $i\" }")
    print(f"Got {len(result['output'])} lines")
    print(f"More available: {result['more_available']}")

    if result['more_available']:
        print("\nGetting more output...")
        more_result = client.more()
        print(f"Got {len(more_result['output'])} more lines")
        print(f"Still more available: {more_result['more_available']}")

    print("\n✓ All examples completed successfully!")


if __name__ == "__main__":
    try:
        main()
    except requests.exceptions.ConnectionError:
        print("Error: Cannot connect to slopdrop server", file=sys.stderr)
        print("Make sure to start the server first:", file=sys.stderr)
        print("  ./target/release/slopdrop --web", file=sys.stderr)
        sys.exit(1)
    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)

File: examples/web_api_client.js
#!/usr/bin/env node
/**
 * Example Node.js client for slopdrop Web API
 * Demonstrates how to interact with slopdrop from JavaScript
 *
 * Usage:
 *   node examples/web_api_client.js
 */

class SlopdropClient {
    constructor(baseUrl = 'http://127.0.0.1:8080') {
        this.baseUrl = baseUrl;
    }

    async eval(code, isAdmin = false) {
        const response = await fetch(`${this.baseUrl}/api/eval`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ code, is_admin: isAdmin })
        });

        if (!response.ok) {
            throw new Error(`HTTP ${response.status}: ${await response.text()}`);
        }

        return await response.json();
    }

    async more() {
        const response = await fetch(`${this.baseUrl}/api/more`);

        if (!response.ok) {
            throw new Error(`HTTP ${response.status}: ${await response.text()}`);
        }

        return await response.json();
    }

    async history(limit = 10) {
        const response = await fetch(`${this.baseUrl}/api/history?limit=${limit}`);

        if (!response.ok) {
            throw new Error(`HTTP ${response.status}: ${await response.text()}`);
        }

        return await response.json();
    }

    async rollback(commitHash) {
        const response = await fetch(`${this.baseUrl}/api/rollback`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ commit_hash: commitHash })
        });

        if (!response.ok) {
            throw new Error(`HTTP ${response.status}: ${await response.text()}`);
        }

        return await response.json();
    }

    async health() {
        const response = await fetch(`${this.baseUrl}/api/health`);

        if (!response.ok) {
            throw new Error(`HTTP ${response.status}: ${await response.text()}`);
        }

        return await response.json();
    }
}

async function main() {
    const client = new SlopdropClient();

    try {
        // Check health
        console.log('Checking server health...');
        const health = await client.health();
        console.log(`✓ Server is ${health.status}\n`);

        // Simple evaluation
        console.log('Evaluating: expr {1 + 1}');
        let result = await client.eval('expr {1 + 1}');
        console.log(`Result: ${result.output.join('\n')}`);
        console.log(`Error: ${result.is_error}\n`);

        // Set a variable
        console.log('Evaluating: set myvar "Hello from JavaScript!"');
        result = await client.eval('set myvar "Hello from JavaScript!"');
        console.log(`Result: ${result.output.join('\n')}\n`);

        // Read the variable
        console.log('Evaluating: set myvar');
        result = await client.eval('set myvar');
        console.log(`Result: ${result.output.join('\n')}\n`);

        // Define a procedure
        console.log('Defining fibonacci procedure...');
        const fibCode = `
proc fibonacci {n} {
    if {$n <= 1} {
        return $n
    }
    expr {[fibonacci [expr {$n - 1}]] + [fibonacci [expr {$n - 2}]]}
}
        `.trim();
        result = await client.eval(fibCode, true);
        console.log(`Result: ${result.output.join('\n')}\n`);

        // Call the procedure
        console.log('Evaluating: fibonacci 10');
        result = await client.eval('fibonacci 10');
        console.log(`Result: ${result.output.join('\n')}\n`);

        // Get history
        console.log('Getting git history...');
        const history = await client.history(5);
        console.log('Recent commits:');
        for (const commit of history.history) {
            console.log(`  ${commit.commit_id.substring(0, 8)} - ${commit.author} - ${commit.message}`);
        }
        console.log();

        // Test pagination
        console.log('Testing pagination with large output...');
        result = await client.eval('for {set i 0} {$i < 50} {incr i} { puts "Line $i" }');
        console.log(`Got ${result.output.length} lines`);
        console.log(`More available: ${result.more_available}`);

        if (result.more_available) {
            console.log('\nGetting more output...');
            const moreResult = await client.more();
            console.log(`Got ${moreResult.output.length} more lines`);
            console.log(`Still more available: ${moreResult.more_available}`);
        }

        console.log('\n✓ All examples completed successfully!');

    } catch (error) {
        if (error.cause?.code === 'ECONNREFUSED') {
            console.error('Error: Cannot connect to slopdrop server');
            console.error('Make sure to start the server first:');
            console.error('  ./target/release/slopdrop --web');
            process.exit(1);
        }
        throw error;
    }
}

main().catch(error => {
    console.error(`Error: ${error.message}`);
    process.exit(1);
});
